futures-util/src/macros/mod.rs

```diff
@@ -1,2 +1,12 @@
 #[macro_use]
 mod poll;
+
+macro_rules! cfg_target_has_atomic {
```

Comment:
There are a bunch of spots in e.g. `futures-util/src/stream/mod.rs` that seem like they could benefit greatly from this-- or does that not work out because they're trait items?

Reply:
> -- or does that not work out because they're trait items?

Yeah, it doesn't work well with trait items... [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=dd94d620920242baac820f0120e915bc)

Reply:
Also, we should be able to work around this problem by using procedural macros, but in that case, I would like to compare with the approach using `proc_macro_attribute`.

1. `proc_macro`

```rust
// definition
#[proc_macro]
pub fn cfg_target_has_atomic(input: TokenStream) -> TokenStream {
    let mut file: File = parse_macro_input!(input);
    let attr = parse_quote! {
       #[cfg_attr(
            feature = "cfg-target-has-atomic",
            cfg(all(target_has_atomic = "cas", target_has_atomic = "ptr"))
        )]
    };
    file.items.iter_mut().for_each(|item| {
        match item {
            Item::Fn(ItemFn { attrs, .. }) 
            | Item::Use(ItemUse { attrs, .. }) 
            | Item::Mod(ItemMod { attrs, .. }) => attrs.push(attr.clone()),
            _ => {}        
        }
    });
    TokenStream::from(file.into_token_stream())
}

// usage
cfg_target_has_atomic! {
    #[cfg(feature = "alloc")]
    mod buffer_unordered;
    #[cfg(feature = "alloc")]
    pub use self::buffer_unordered::BufferUnordered;
}
```

2. `proc_macro_attribute`

```rust
// definition
#[proc_macro_attribute]
pub fn cfg_target_has_atomic(_args: TokenStream, input: TokenStream) -> TokenStream {
    let mut tts = TokenStream::from(quote! {
       #[cfg_attr(
            feature = "cfg-target-has-atomic",
            cfg(all(target_has_atomic = "cas", target_has_atomic = "ptr"))
        )]
    });
    tts.extend(input);
    tts
}

// usage
#[cfg_target_has_atomic]
#[cfg(feature = "alloc")]
mod buffer_unordered;
#[cfg_target_has_atomic]
#[cfg(feature = "alloc")]
pub use self::buffer_unordered::BufferUnordered;
```

`1` is especially useful when many items are in one place, but the indent is deeper. 
(I have no strong opinion on which one to use.)

Reply:
Yeah, that'd work-- I don't have particularly strong thoughts either way. We can leave that for a future PR, though, since it doesn't seem like the current situation is too bad (just some extra boilerplate in a few spots).